### PR TITLE
MPI testing

### PR DESCRIPTION
**Pull request by claireh93**
Further testing to MPI to check answers are the same under different numbers of processes.
Add `gather_abundance()` to pull all abundances to the root node in the correct order.

### CODE REVIEW

**Comment by claireh93** [comment]
@richardreeve I think this is ready for you to have a look at now

File: test/test_MPI.jl [hunk 1]
```diff
@@ -60,5 +61,43 @@ using MPI
...
+    @test sum(eco.abundances.cols_vector) == sum(eco.abundances.rows_matrix)
+
+    # Gather abundances and check against rows matrix - should be same for 1 process
+    abuns = gather_abundance(eco)
+    @test abuns == eco.abundances.rows_matrix
+
     MPI.Finalize()
 end
+
+@testset "" begin
```
Comment (richardreeve) [inline test/test_MPI.jl:81]:
```suggestion
@testset "mpirun" begin
```

File: test/test_MPI.jl [hunk 2]
```diff
@@ -3,6 +3,7 @@ using EcoSISTEM.Units
 using Unitful, Unitful.DefaultSymbols
 using Distributions
 using MPI
+using JLD
 
 @testset "MPI" begin
```
Comment (richardreeve) [inline test/test_MPI.jl:8]: I think this is only running with one process?

File: test/test_MPI.jl [hunk 3]
```diff
@@ -60,5 +61,43 @@ using MPI
...
+
+    # Compare 1 thread 4 processes vs. 4 threads 1 process vs. 2 threads 2 processes
+    ENV["JULIA_NUM_THREADS"] = 1
+    mpiexec(cmd -> run(`$cmd -n 4 julia SmallMPItest.jl`));
+
+    ENV["JULIA_NUM_THREADS"] = 2
+    mpiexec(cmd -> run(`$cmd -n 2 julia SmallMPItest.jl`));
+
+    ENV["JULIA_NUM_THREADS"] = 4
+    mpiexec(cmd -> run(`$cmd -n 1 julia SmallMPItest.jl`));
```
Comment (richardreeve) [inline test/test_MPI.jl:93]: This is the bit I was referring to earlier that I think doesn't get picked up by the code coverage tool?

File: test/test_MPI.jl [hunk 4]
```diff
@@ -60,5 +61,43 @@ using MPI
...
+
+    ## All answers should be the same
+    abuns1thread = load("data/Test_abuns1.jld", "abuns")
+    abuns2thread = load("data/Test_abuns2.jld", "abuns")
+    abuns4thread = load("data/Test_abuns4.jld", "abuns")
+
+    @test abuns1thread == abuns2thread == abuns4thread
+    # Clean up outputs
+    rm("data", recursive = true)
+end
```
Comment (richardreeve) [inline test/test_MPI.jl:103]: Missing newline at end of file...
```suggestion
end

```

**Review by richardreeve — CHANGES_REQUESTED** [review]
Generally looks okay, and maybe we can ignore the code coverage / testing problem for now, but if you can see a way of fixing it, that would be great...

File: test/test_MPI.jl (hunk 3 above)
Comment (claireh93) [reply test/test_MPI.jl:93]: Hmm I can't see it not being picked up on the coverage tool - I don't think `SmallMPItest.jl` has any code in it that isn't tested elsewhere anyway so maybe it's fine? I can't see an obvious way to do it on MPI.jl anyway

File: test/test_MPI.jl (hunk 2 above)
Comment (claireh93) [reply test/test_MPI.jl:8]: Yes it is - but maybe that's fine, if we want to ensure that codecov picks up some of it?

File: test/test_MPI.jl (hunk 3 above)
Comment (claireh93) [reply test/test_MPI.jl:93]: Oh sorry - I read your previous comment now about the VBuffer line!

File: src/MPILandscape.jl [hunk 5]
```diff
@@ -33,6 +34,10 @@ mutable struct MPIGridLandscape{RA <: Base.ReshapedArray, NT <: NamedTuple}
       reshape(view(cols_vector, (sppindices[i] + 1) : sppindices[i + 1]),
               Int64(sppcounts[i]), Int64(sccounts[rank + 1]))
     end
+    reshaped_cols_vector = vcat(map(1:length(sccounts)) do i
+      reshape(view(cols_vector, (sppindices[i] + 1) : sppindices[i + 1]),
+              Int64(sppcounts[i]), Int64(sccounts[rank + 1]))
+    end ...)[1:end]
```
Comment (richardreeve) [inline src/MPILandscape.jl:40]: So you create this `reshaped_cols_vector` here, but you don't seem to use it... is it supposed to be used here:

https://github.com/boydorr/EcoSISTEM.jl/blob/09e21e946e03ce9e9ad1cc862319df94723a7fe3/src/MPIEcosystem.jl#L110

If so (it looks a bit different?) you haven't edited the code there.

Comment (claireh93) [reply src/MPILandscape.jl:40]: Oh dear, no wonder it worked! Sorry, I'm having a bit of trouble with this one - I can't get it to sync up properly

**Comment by richardreeve** [comment]
@claireh93 I see your problem here. I hadn't realised this never worked before. For the time being your code works, so let's stick with it and remove the allocation later. It's not easy at all to see how to fix this. I think fundamentally we may need something like a view into an array - possibly in the root process - or distributed calculation of the diversities and abundances...

So, assuming this works I think we should just merge it as is.

**Review by richardreeve — APPROVED** [review]
Looks good to me now.

**Comment by richardreeve** [comment]
PS Another issue I noticed is that (as I'm sure you've noticed) JLD is broken and generates warnings - we need to move away from it I guess. I don't know if the solution is to use the data pipeline or to switch to JLD2, but I'll create an issue about that too.

**Pull request merged**